### ISSUE TITLE
Jaeger Query can OOM when retrieving large traces

### ISSUE THREAD

**Issue by vprithvi** [jaegertracing/jaeger#1051]
## Requirement - what kind of business use case are you trying to solve?
- Retrieve large traces
- Be resilient against bad instrumentation using same traceID for all traces

## Problem - what in Jaeger blocks you from solving the requirement?
Jaeger Query OOMs on retrieval of large traces on Cassandra. 
If someone is crafty, they can easily create a trace with millions of spans, and attempt to retrieve it to systematically bring down all jaeger-query instances. 

## Any open questions to address
- Does this affect ES as well?

**Comment by pavolloffay** [comment]
Related issue to ES, although its related to overloading ES not query https://github.com/jaegertracing/jaeger/issues/960

**Comment by jpkrohling** [comment]
We probably want to add this scenario to our test cases

cc @jkandasa

**Comment by yurishkuro** [comment]
We may consider adding LIMIT to Cassandra queries by trace ID.

**Comment by pavolloffay** [comment]
Limit on the number of spans in a trace?

**Comment by yurishkuro** [comment]
Well there are two separate things. We're seeing some traces in production with many millions of spans so we want to implement a limit on ingestion. But separately, there can be a limit on reading side to avoid loading too much even if that many spans were saved.

**Comment by annanay25** [comment]
The olivere/elastic library supports sending a range query using `From(0)` and `Size(100)` to fetch a predefined number of documents - for ex: https://github.com/olivere/elastic/blob/release-branch.v6/example_test.go#L172

Would it make sense to pass this range as a param in MultiSearch here - https://github.com/jaegertracing/jaeger/blob/master/plugin/storage/es/spanstore/reader.go#L256 ?

**Comment by pavolloffay** [comment]
Perhaps yes, I am not 100% percent sure how the current search works. Would you like to submit a patch?

**Comment by annanay25** [comment]
Sure, I'll give it a shot soon.

**Comment by annanay25** [comment]
@pavolloffay could you review this? https://github.com/annanay25/jaeger/commit/2d167b86a14f1a88a6e7ed7cee84399f8403ba88
Reduced the default fetch limit from 10,000 to 1,000. And added a `FetchNext` function to implement scroll.

**Comment by yurishkuro** [comment]
I think the simplest fix is what is proposed in the description of this ticket. It's not ideal since it will truncate the data, but we can revisit that later when/if we support a streaming API from the query service. At least this fix is fairly straightforward and will protect the query svc from OOM. Most of the analytics will choke on a trace with >100k spans anyway. The limits will be an optional configuration.

**Comment by ZigmunD** [comment]
@yurishkuro I agree. Maybe should just mark trace in web interface that it is truncated due span limit.

**Comment by annanay25** [comment]
@yurishkuro truncated search results implemented for ES and Cassandra - https://github.com/annanay25/jaeger/commit/9eb7590176383f4f6e95c7194a761b1952f8d024

We were waiting on a test case for this? Is someone working on that?

**Comment by yurishkuro** [comment]
Not afaik

**Comment by mohit-chawla** [comment]
Not 100% related but i have a question: 

**Edge case:** For a use case, i need to use in mem storage, was wondering about edge case where one huge trace takes up whole memory, will this cause any OOM errors or will the trace be discard? 
- do we have a max trace size available ?
- how can i increase the mem allocated ? 

cc: @annanay25

**Comment by Sreevani871** [comment]
Encountering OOM issues while fetching traces from jaeger-UI. 
**Details :** 
Backend Storage: Elasticsearch
Instance RAM: 4 GB(~3.7 GB Free Memory)
Instance vCPU:2

We are encountering OOM kill upon fetching 20 traces(each trace consists of ~20K spans) from UI. This is causing frequent process restarts.
While going through code, observed jaeger-query service FindTraces API is fetching all spans of a resulted traces in this call itself. 
Would it be more efficient if it is feasible to fetch only top traceID's and related stats info to display in UI by using ES count queries and all spans of a trace can be queried on click on particular trace. In this case also OOM issue will come if a single trace has large set of spans. To avoid this can we just load default no of spans for a trace initially and providing an option like **load more** in UI to scroll further set of spans till the trace has no more spans left.

**Comment by yurishkuro** [comment]
@Sreevani871 yes, we even had tickets somewhere for that, but no volunteers to implement it.

**Comment by Sreevani871** [comment]
@yurishkuro Any specific reason for fetching all info in **FindTraces** call itself from UI?

**Comment by yurishkuro** [comment]
If you mean the internal implementation of FindTraces, then yes, the reason is because there is no summary information about the full trace available anywhere else, since collectors receive spans one at a time. Storage is the only place that can provide a full trace, but you need to read that full trace in order to extract summary. It's possible to build a post-processing, but it will require additional infrastructure/coordination.

**Comment by Sreevani871** [comment]
> @Sreevani871 yes, we even had tickets somewhere for that, but no volunteers to implement it.

Can you share the tickets will have a look?

**Comment by yurishkuro** [comment]
There is context here: https://github.com/jaegertracing/jaeger-ui/issues/247

**Comment by Sreevani871** [comment]
@yurishkuro Any progress on this?

**Comment by Sreevani871** [comment]
To overcome frequent OOM kills of jaeger-query service (Due to traces having a large number of spans), we need to avoid fetching of all spans data of respective TopN traces. Here listing out the approaches we thought of to solve the issue. 
we are using elasticsearch as span storage. Please review and share your feedback

**Approach - I:**
 Fetch summary of TopN traces instead of fetching all raw span data of respective traces from the backend
 Returning the summary instead of all raw span data to UI avoids the summary generation step at the UI side which intern saves UI page load time.
**Changes Required at Backend:**
 - Introduce a new API: /summary to fetch a summary of traces. 
 - Need to add code flow for /summary API which involves changes at querier and span reader interface which will enforce all span storage types to implement summary method at the reader side. 
 - The API returns response as **map[TraceID]TraceSummary{}**

```
type TraceSummary  {
    ServiceName string.        //  service name of the old span of trace
    OperationName string    // operation name of the old span of trace
    StartTime string             // start time of the old span of a trace.
    Duration time.Duration // duration of the old span of a trace.
    SpanCount int               // total span count of a trace
    SpanCountByService map[string]int
    ErrorSpanCount int // error span count of a trace
    SpanID string
    }
```
Here old span of trace refers to parent span ideally. To avoid cases like parent span missing or parent span not indexed at the time fo search, considering fetching old span(sort by startTime) of a trace available at the time of the search.

**Code Changes Required at UI:**
- UI has to call a new API to get the summary of traces data. 
- Make use of summary data returned from the backend to generate the FindTraces page.
- Since UI does not have full span data of a trace, On click on trace, UI should call the backed for full span details.

**Pros:**

- The memory usage for this is approach is very minimal and no.of the objects returned from the backend to UI is equal to no.of traces we fetch and each trace object carries the only summary.
- Memory usage is better than Approach-2

**Cons:**
- Requires code changes at both backend and frontend. Compare to Approach-2 UI requires considerable code changes to adopt this approach.

**Note**: FindTraces page has a _Deep Dependency Graph_ option, Since we are avoiding fetching all span data in the FindTraces API call, Need to have a separate API which will fetch the required fields of spans for traces to construct DDG which will overlap with our idea of Approach-2.


**Approach - II:**
Analyze the required fieldset of span to compute the FindTraces UI page without modifying the existing API response data model.
Observed following fieldset of the span are being used to generate the FindTraces list view and DDG.
1. ServiceName
2. OperationName
3. References
4. Tags.Span.Kind
5. StartTime
6. Duration

**Changes Required At Backend:**
- Change the ES query to fetch only selective fields from the source.
- The following specifies data flow from ES to UI back.
-  FindTraces() method returns []jaeger.model.Trace
        `ES -> []byte -> MultiSearchResults{} -> es.dbmodel.Span -> jaeger.model.Span -> jaeger.model.Trace`
        - ConvertToUI model
           ` []jager.model.Trace -> []json.Trace `
- In the above flow data transformation(serialization/deserialization) steps are consuming 2 to 3X extra memory of actual data. Confirmed from heap profile alloc_space data.
<img width="1119" alt="Screen Shot 2020-10-12 at 2 26 20 PM" src="https://user-images.githubusercontent.com/20114703/95726953-2f69f400-0c97-11eb-8c23-925d6962e325.png">

- To avoid this extra memory usage we can reduce these steps and will have the following code flow 
        ES ->[]byte -> MultiSearchResults{} -> []json.Trace.

**Changes Required At Frontend:**
    - No considerable code changes required at the UI side expect the API call if we introduce a new API for this approach.
**Pros:**
- Only changes required at the backend side. No changes required at the frontend side.
- It will reduce the memory consumption by x percentage(since we are avoiding fetching unnecessary fieldset of a span).

**Cons:**
- It will reduce the memory by x percentage compared to existing code, but no the objects returned from backed to UI not reduced here. 
Ex: Fetching 20 traces with each trace has 20K spans resulting in 400K objects.
Approach-1 is comparatively better in terms of memory usage reduction and the UI page load time since we reduced UI operations of building a summary from data returned from the backend.

**Comment by yurishkuro** [comment]
> Note: FindTraces page has a Deep Dependency Graph option...

That's a very astute observation, since DDG is built entirely by the UI and requires loading all spans of the trace (but not all data in the spans).

A more flexible variation of approach 2 is to build support for GraphQL. It's going to be difficult to support partial field retrieval at the storage level (some storage implementations may only allow retrieving the whole trace as a blob), so the whole data is still going to be loaded in memory from the database. The memory savings can come from refactoring the storage API to be more like stream, i.e. instead of synchronous:
```go
FindTraces(query) ([]*model.Span, error)
```
implement the streaming version (incidentally, the grpc-plugin storage API already uses streaming under the hood):
```go
FindTracesStreaming(query, handler func(*model.Span) error) error
```
The handler function can immediately convert model.Span to JSON output object, AND trim it down by filtering only the fields required by GraphQL query. 

This will not reduce the number of memory allocations, but will allow query service not to hold onto every single span of the whole result set.

Later this can be further extended into approach 1, which will reduce memory requirements even further (but DDG from search will likely require a second query, which is probably ok since not every search is used to display DDG).

**Comment by Sreevani871** [comment]
Using _json.Unmarshal()_ func instead of _json.Decode()_ func here  https://github.com/jaegertracing/jaeger/blob/master/plugin/storage/es/spanstore/reader.go#L233 shows improvement on memory usage because json.Decode() func is buffering entire json value in memory again. Since here the data passed to unmarshalJSON() method is already in memory so using json.Unmarshal() is improving performance in terms of memory usage.
Benchmarking Results:
```

goos: darwin
goarch: amd64
BenchmarkUnmarshal-8     2251400               527 ns/op             352 B/op          5 allocs/op
BenchmarkDecoder-8       1470145               812 ns/op            1096 B/op          8 allocs/op

```
```
package test

import (
	"bytes"
	"testing"
	"fmt"
	"encoding/json"
)

type JSON struct {
	a []int
	b string
	c float64
	d []byte
}

var j = JSON{
	a: []int{1,2,4,5,566,9},
	b: "Testing Data",
	c: 145.0,
	d: []byte("Testing Data"),
}
func BenchmarkUnmarshal(b *testing.B){
	for n := 0; n < b.N; n++ {
		unmarshalJson(marshalJson(j))
	}
}

func BenchmarkDecoder(b *testing.B){
	for n := 0; n < b.N; n++ {
		decodeJson(marshalJson(j))
	}
}

func marshalJson(j JSON) []byte{
	v,err:=json.Marshal(j)
	if err !=nil {
		fmt.Println(err)
		return nil
	}
	return v
}
func unmarshalJson(data []byte) {
	var j =JSON{}
	err:=json.Unmarshal(data, &j)
	if err!=nil {
		fmt.Println(err)
	}
}

func decodeJson(data []byte) {
	var j = JSON{}
	d := json.NewDecoder(bytes.NewReader(data))
	d.UseNumber()
	if err := d.Decode(&j); err != nil {
		fmt.Println(err)
	}
}

**Comment by yurishkuro** [comment]
Interesting query benchmark numbers from [a post on Slack](https://cloud-native.slack.com/archives/CGG7NFUJ3/p1691431445212979):
* 1 trace with approximately 1700 spans returns in 1.5 seconds. (5 unique ES queries (+ 2 ES queries added per any additional trace/limit) all complete in under a second)
* 10 traces, each with approximately 1700 spans, return in 5 seconds. Memory: 150MB
* 100 traces, each with approximately 1700 spans, return in 30 seconds. Memory: 700MB - 1.5GB (203 unique ES queries, all of which complete in under 3 seconds)

**Comment by jkowall** [comment]
One possible solution here would be to limit the number of spans returned when pulling up a trace to 50k and allow this to be configurable.

**Comment by yurishkuro** [comment]
Many of the issues discussed in this issue will be alleviated by Storage v2 API (#5079) which will support streaming of the results. It will not address the issue if a single trace has millions of spans since it still needs to be loaded by the query service in order to apply adjustments and conversion, but the query service will be in control of pulling as much data as it wants to it can put a limit on the number of spans it's willing to receive and terminate the load after that.

**Comment by iypetrov** [comment]
Can we close this ticket since https://github.com/jaegertracing/jaeger/issues/5079 has already been merged?

**Comment by yurishkuro** [comment]
I think it makes sense to at least add a configuration to query for the max number of spans it should be trying to read from storage. I added #7495.

**Comment by AnmolxSingh** [comment]
@yurishkuro 
Till now what i have understood the issue of OOM when retrieving large traces has been solved for server side what I believe by using streaming introduced in Storage v2 API. But what about the user side ?? OOM issue can occur when displaying trace with all those spans together at UI side in browser.

> One possible solution here would be to limit the number of spans returned when pulling up a trace to 50k and allow this to be configurable.

Could this have not been the solution with displaying error something like could not retrieve all the spans?

From my point of view someone is rarely going to scan more than 50k spans in a trace.

**Comment by yurishkuro** [comment]
From perspective the main focus should be system stability. The server must not crash because of large trace. Some adjustments post query require full trace in memory (eg clock skew adjustment) but the server may opt to bypass those if the trace is too large. Similarly, the UI can warn the user that the trace is too large, but if the UI can still handle it without OOMing it should give the option to do so, possibly with degraded functionality like not computing critical path. Putting a hard max size block is the easiest but the least creative solution, but I'm ok with it if it's opt-in for site operator.

**Comment by AnmolxSingh** [comment]
Yes agreed. Putting a hard max size block seems the least creative solution. But the issue #7495 does not appear to be opt-in for site operator it looks more like a temporary hard fix.

**Comment by yurishkuro** [comment]
Yes, but it's better to have some protection lever than none.

**Comment by feldentm-SAP** [comment]
If a max trace size limit has been implemented, isn't the missing part here documenting that parameter in the performance tuning documentation?
